content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+
+
+## Reference Solution
+
+Some test suites require a reference solution to generate the expected value(s) and compare them to those returned by the user solution. While such practice is common and sometimes necessary, there are things which have to be handled carefully:
+
+- **Avoid using a reference solution if at all possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary. Most of the time, test cases can be generated with the answer known upfront which eliminates an entire class of potential problems associated with reference solutions:
+  - Extra time spent on running a reference solution
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution affecting the input passed to the reference solution
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global function.
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
```

Comment:
"make sure" -> it's actually not imperative, but strongly recommended. Sometimes it's hard to achieve...

=> soften the sentence?

Reply:
Maybe I could just remove it? It feels somewhat reduntand with the first bullet point, so they could be combined, I think...

```suggestion
```